src/pdb/mod.rs

```diff
@@ -278,61 +283,87 @@ pub struct Page {
     /// Number of rows in this page.
     ///
     /// **Note:** This is a virtual field and not actually read from the file.
-    #[br(temp)]
-    #[br(calc = if num_rows_large > num_rows_small.into() && num_rows_large != 0x1fff { num_rows_large } else { num_rows_small.into() })]
+    #[br(temp, calc = Self::calculate_num_rows(num_rows_small, num_rows_large))]
     num_rows: u16,
+    /// Number of rows groups in this page.
+    #[br(temp, calc = num_rows.div_ceil(RowGroup::MAX_ROW_COUNT as u16))]
+    num_row_groups: u16,
     /// The offset at which the row data for this page are located.
     ///
     /// **Note:** This is a virtual field and not actually read from the file.
-    #[br(temp)]
-    #[br(calc = page_index.offset(page_size) + u64::from(Self::HEADER_SIZE))]
+    #[br(temp, calc = page_start_pos + u64::from(Self::HEADER_SIZE))]
     page_heap_offset: u64,
     /// Row groups belonging to this page.
-    #[br(seek_before(SeekFrom::Current(i64::from(page_size) - i64::from(Self::HEADER_SIZE))), restore_position)]
-    #[br(parse_with = Self::parse_row_groups, args(page_type, page_heap_offset, num_rows, page_flags))]
+    #[br(seek_before(SeekFrom::Current(Self::heap_padding_size(page_size, num_row_groups).into())))]
+    #[br(args {count: num_row_groups.into(), inner: (page_type, page_heap_offset)})]
+    #[br(if(num_row_groups > 0 && page_flags.page_has_data()))]
     pub row_groups: Vec<RowGroup>,
 }
 
-impl Page {
-    /// Size of the page header in bytes.
-    pub const HEADER_SIZE: u32 = 0x28;
+// #[bw(little)] on #[binread] types does not seem to work so we manually define the endianness here.
+impl binrw::meta::WriteEndian for Page {
+    const ENDIAN: binrw::meta::EndianKind = binrw::meta::EndianKind::Endian(Endian::Little);
+}
 
-    /// Parse the row groups at the end of the page.
-    fn parse_row_groups<R: Read + Seek>(
-        reader: &mut R,
-        _: Endian,
-        args: (PageType, u64, u16, PageFlags),
-    ) -> BinResult<Vec<RowGroup>> {
-        let endian = Endian::Little;
+impl BinWrite for Page {
+    type Args<'a> = (u32,); // page_size
 
-        let (page_type, page_heap_offset, num_rows, page_flags) = args;
-        if num_rows == 0 || !page_flags.page_has_data() {
-            return Ok(vec![]);
-        }
+    fn write_options<W: Write + Seek>(
+        &self,
+        writer: &mut W,
+        endian: Endian,
+        args: Self::Args<'_>,
+    ) -> BinResult<()> {
+        let (page_size,) = args;
+        let page_offset = writer.stream_position()?;
+
```

Comment:
this has gotten quite simple/repetitive. Do you want to take a stab at doing a similar transformation as I did with by removing the manual `BinRead` impls? It should be quite straightforward by using [`#[write_with]`](https://docs.rs/binrw/latest/binrw/docs/attribute/index.html#custom-parserswriters) on `row_groups`